### PR TITLE
Add resource management as part of the kore shutdown process.

### PR DESCRIPTION
**Pull request by oneswig**
Hi Joris - 

What a neat codebase you have created.  I added some management of buffers and file descriptors on shutdown.  This was to get Kore more clean when running under Valgrind.  I haven't touched anything related to memory management during the config parsing and validation process - from the comments it looks like you're planning to rework this at some point anyway?

Thanks and best wishes
Stig

### CODE REVIEW

File: src/mem.c [hunk 1]
```diff
@@ -89,12 +89,19 @@ kore_realloc(void *ptr, size_t len)
 void *
 kore_calloc(size_t memb, size_t len)
 {
-	if (memb == 0 || len == 0)
+	void *result;
+	const size_t nbytes = memb * len;
+
+	if (nbytes == 0)
 		fatal("kore_calloc(): zero size");
-	if (SIZE_MAX / memb < len)
```
Comment (jorisvink) [inline src/mem.c]: Hi,

Thanks for this, I never got around writing it and wasn't particularly happy with valgrind being angry at me :-)

Why did you remove the overflow check in kore_calloc() though? It is now susceptible to an integer overflow. The check you introduced is just plain wrong.

Rest looks promising. Although I am not a big fan of the fini naming convention.

.joris

Comment (oneswig) [inline src/mem.c]: Hi Joris - 

Ah, I’ve looked again and now I understand your intent with potential size_t overflow of memb * len.  Thanks for pointing that out.

More generally I’m curious about why you’re reimplementing malloc, calloc, realloc in this way?  The realloc in particular is going to be slower than the glibc implementation, and it gets used on the socket receive path.

Best wishes,
Stig

**Comment by jorisvink** [comment]
Can you rename the _fini stuff to _cleanup instead? I'm all game for the other changes.

**Comment by jorisvink** [comment]
Still broken under BSD / OSX due to not having changed bsd.c :)

.joris

**Comment by jorisvink** [comment]
Great, I noticed you amended the BSD stuff. It seems to be working on all platforms now.

A few notes:
* http_cleanup() needs to cleanup the http_body_path pool (was only recently introduced)
* the module handlers and validators are not cleaned up properly.

Other then those I'm quite happy with this PR! Having this merged in will make it easier going forward for finding actual leaks and whatnot.

**Comment by oneswig** [comment]
I think these commits cover the additional points you raised.

There are still more leaks across the codebase, but this is now going well beyond the scope of my use of Kore and I do not have unit tests or an application test case to verify.

Enjoy,
Stig

**Comment by jorisvink** [comment]
Leaks meaning resources not cleaned up during shutdown.
Not actual resource leaks during runtime, right?

**Comment by oneswig** [comment]
Correct - at least that’s my understanding of it.

**Comment by jorisvink** [comment]
OK.

I'll make sure this gets merged tonight, thanks for the hard work.

**Comment by oneswig** [comment]
No worries Joris, you’re welcome :-)
